Handle various redirects

adamhughes86 opened the pull request:
Adds the redirect.html as an override which is detailed in the [jekyll-redirect-to repo](https://github.com/jekyll/jekyll-redirect-from/blob/master/lib/jekyll-redirect-from/redirect.html). This has then been modified to grab the location hash and append it to the redirect url.

Tested locally using [https://playbook.dxw.com/guides/support-and-on-call.html#claiming-toil-for-out-of-hours-alerts](https://playbook.dxw.com/guides/support-and-on-call.html#claiming-toil-for-out-of-hours-alerts). Not sure if it covers all hash redirect issues but seems to improve some. 

Resolves #1032

snim2 commented:
Thanks for this, @adamhughes86

snim2 commented:
@yndajas I'm not sure whether you want to leave that issue up for a bit longer, just to check that this covers everything?

yndajas commented on src/_layouts/redirect.html:
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
+<html lang="en-US">
```
Shall we keep this as `en-GB` in line with `default.html`?

adamhughes86 replied:
Yeah that makes sense. It's in our codebase now should so match the standards. I'll update this template based on the default.html

yndajas commented on src/_layouts/redirect.html:
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
+<html lang="en-US">
+  <meta charset="utf-8">
+  <title>Redirecting&hellip;</title>
+  <link rel="canonical" href="{{ page.redirect.to }}">
+  <script>location = "{{ page.redirect.to }}" + location.hash</script>
+  <meta http-equiv="refresh" content="0; url={{ page.redirect.to }}">
+  <meta name="robots" content="noindex">
+  <h1>Redirecting&hellip;</h1>
+  <a href="{{ page.redirect.to }}">Click here if you are not redirected.</a>
```
Guess it doesn't matter, but is it worth including the conventional `head` and `body` tags?

adamhughes86 replied:
I didn't even notice they weren't in there. I'll update to match the standards in the default.html

yndajas reviewed (approved):
Nice! Couple of trivial comments, but this seems to work great

adamhughes86 commented:
@yndajas @snim2 

I've added more onto this pr related to the redirects. I believe this should cover all requirements now but it could do with a good review as I've not used netlify before.

- Created reusable redirect script and applied to homepage and 404 file.
- Created redirect manager for netlify.
- Added redirect file that netlify will append to with first couple of redirects

yndajas commented on src/assets/js/redirect.js:
```diff
@@ -0,0 +1,12 @@
+var fullPath = location.pathname + location.hash
+
+fetch('/redirects.json')
+  .then((response) => response.json())
+  .then((json) => {
+  var redirectArray = json.redirects
+  var redirectResult = redirectArray.find(item => item.from === fullPath);
+
+  if (redirectResult) {
+    location = redirectResult.to
+  }
+});
```
Since #984 we've adopted [Standard](https://standardjs.com) on this repo. It's not yet integrated with the CI (see https://github.com/dxw/playbook/issues/1031), so non-conformant code won't produce errors, but it would be good to keep new JS code in line if possible. Using an auto-formatting extension in VSCode, it fixes this to the following to align with Standard:

```
const fullPath = location.pathname + location.hash

fetch('/redirects.json')
  .then((response) => response.json())
  .then((json) => {
    const redirectArray = json.redirects
    const redirectResult = redirectArray.find(item => item.from === fullPath)

    if (redirectResult) {
      location = redirectResult.to
    }
  })
```

Would you be able to update the code to match?

adamhughes86 replied:
Yeah no problem. I'll sort that now

adamhughes86 replied:
In reference to the points above:

1. I will add some details to the CONTRIBUTING.md file.
2. I think as long as it's documented we should keep both for now and only use the redirect manager if required. We could change this down the line if we think it's worth it.
3. I'll add an extra ticket. We could probably put a simple bit of JS in there to hide the 404 container until the redirect script is complete
4. That would be a good feature. I'll add it to the issues for now

yndajas reviewed (approved):
This works great! Tested locally and the redirects worked as expected, and I was able to edit the list of redirects via the Netlify CMS 🎉

I've made one comment on the script style. A few of other things came to mind, but nothing that should be a blocker if you want to get this merged (can always write these up as separate issues for future work):
1. Could we update `CONTRIBUTING.md` to explain how redirects work? We now have two different ways of creating redirects:
    - using the `jekyll-redirect-from` gem, adding a `redirect_from` entry to a page's front matter - this works for redirecting one page to another, including keeping the same hash fragment thanks to your first commit
    - using the Netlify CMS redirect manager - this works on the homepage and when you get a 404, and is made to plug a specific gap in the above: redirecting from a section of an old page (via a hash fragment) to a new dedicated page
2. Would it be worth wholesale replacing `jekyll-redirect-from` with this new approach? Perhaps it's best to keep the gem, since it's easier to spot/understand a redirect in the front matter of a specific page than in a long list, and just use the redirect manager by exception
3. The 404 page shows briefly before redirects - is it possible to get this script to halt page load/render until complete, so we don't get the quick 404 flash? I can't remember how feasible that is in fairly vanilla JavaScript/HTML
4. A nice to have would be for the JSON and/or Netlify CMS to order the redirects alphabetically. If the list gets long, it might be hard to see if a redirect already exists (although if it does, you probably wouldn't be trying to add it). The list is sortable via drag and drop, but that's reliant on user intervention. As far as I can see, the list widget doesn't support this natively
